Cabal/tests/UnitTests/Distribution/Compat/Graph.hs

```diff
@@ -0,0 +1,88 @@
+{-# LANGUAGE PatternGuards #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+module UnitTests.Distribution.Compat.Graph
+    ( tests
+    , arbitraryGraph
+    ) where
+
+import Distribution.Compat.Graph
+
+import Distribution.Simple.Utils (ordNub)
+
+import qualified Prelude
+import Prelude hiding (null)
+import Test.Tasty
+import Test.Tasty.QuickCheck
+import qualified Data.Set as Set
+import Control.Monad
+import qualified Data.Graph as G
+import Data.Array ((!))
+import Data.Maybe
+import Data.List (sort)
+
+tests :: [TestTree]
+tests =
+    [ testProperty "arbitrary unbroken" (prop_arbitrary_unbroken :: Graph (Node Int ()) -> Bool)
+    , testProperty "nodes consistent" (prop_nodes_consistent :: Graph (Node Int ()) -> Bool)
+    , testProperty "edges consistent" (prop_edges_consistent :: Graph (Node Int ()) -> Property)
+    , testProperty "closure consistent" (prop_closure_consistent :: Graph (Node Int ()) -> Property)
+    ]
+
+-- Our arbitrary instance does not generate broken graphs
+prop_arbitrary_unbroken :: Graph a -> Bool
+prop_arbitrary_unbroken g = Prelude.null (broken g)
+
+-- Every node from 'toList' maps to a vertex which
+-- is present in the constructed graph, and maps back
+-- to a node correctly.
+prop_nodes_consistent :: (Eq a, IsNode a) => Graph a -> Bool
+prop_nodes_consistent g = all p (toList g)
+  where
+    (_, vtn, ktv) = toGraph g
+    p n = case ktv (nodeKey n) of
+            Just v  -> vtn v == n
+            Nothing -> False
+
+-- A non-broken graph has the 'nodeNeighbors' of each node
+-- equal the recorded adjacent edges in the node graph.
```

Comment:
Data.Graph stores adjacency list, so duplicates should be preserved. Although I guess the current code doesn't test it,